backends/graphics/opengl/pipelines/libretro.cpp

```diff
@@ -0,0 +1,541 @@
+
+	// Finally, we need to render the result to the active framebuffer.
+	_applyProjectionChanges = true;
+	setFramebuffer(targetBuffer);
+	_applyProjectionChanges = false;
+
+	ShaderPipeline::activateInternal();
+	ShaderPipeline::drawTexture(*_passes[_passes.size() - 1].target->getTexture(), coordinates, texcoords);
+	ShaderPipeline::deactivateInternal();
+}
+
+void LibRetroPipeline::setProjectionMatrix(const GLfloat *projectionMatrix) {
+	if (_applyProjectionChanges) {
+		ShaderPipeline::setProjectionMatrix(projectionMatrix);
+	}
+}
+
+void LibRetroPipeline::activateInternal() {
+}
+
+void LibRetroPipeline::deactivateInternal() {
+}
+
+void LibRetroPipeline::setOutputSize(uint outputWidth, uint outputHeight) {
+	_outputSizeChanged = (_outputWidth != outputWidth || _outputHeight != outputHeight);
+
+	// Save output dimensions.
+	_outputWidth  = outputWidth;
+	_outputHeight = outputHeight;
+}
+
+bool LibRetroPipeline::open(const Common::FSNode &shaderPreset) {
+	_shaderPreset = LibRetro::parsePreset(shaderPreset);
+	if (!_shaderPreset)
+		return false;
+
+	if (!loadTextures()) {
+		close();
+		return false;
+	}
+
+	if (!loadPasses()) {
+		close();
+		return false;
+	}
+
+	return true;
+}
+
+void LibRetroPipeline::close() {
+	for (TextureArray::size_type i = 0; i < _textures.size(); ++i) {
+		if (_textures[i].textureData) {
+			_textures[i].textureData->free();
+		}
+		delete _textures[i].textureData;
+		delete _textures[i].glTexture;
+	}
+	_textures.clear();
+
+	for (PassArray::size_type i = 0; i < _passes.size(); ++i) {
+		delete _passes[i].shader;
+		delete _passes[i].target;
+	}
+	_passes.clear();
+
+	delete _shaderPreset;
+	_shaderPreset = nullptr;
+}
+
+static Common::FSNode getChildRecursive(const Common::FSNode &basePath, const Common::String &fileName) {
+	Common::FSNode finalPath = basePath;
+	Common::StringTokenizer tok(fileName, "\\/");
+	while (!tok.empty()) {
+		finalPath = finalPath.getChild(tok.nextToken());
+	}
+	return finalPath;
+}
+
+bool LibRetroPipeline::loadTextures() {
+	for (LibRetro::ShaderPreset::TextureArray::const_iterator
+		 i = _shaderPreset->textures.begin(), end = _shaderPreset->textures.end();
+		 i != end; ++i) {
+		Texture texture = loadTexture(getChildRecursive( _shaderPreset->basePath, i->fileName));
+		texture.id = i->id;
+
+		if (!texture.textureData || !texture.glTexture) {
+			return false;
+		}
+
+		texture.glTexture->enableLinearFiltering(i->filteringMode == LibRetro::kFilteringModeLinear);
+		_textures.push_back(texture);
+	}
+	return true;
+}
+
+bool LibRetroPipeline::loadPasses() {
+	for (LibRetro::ShaderPreset::PassArray::const_iterator
+		 i = _shaderPreset->passes.begin(), end = _shaderPreset->passes.end();
+		 i != end; ++i) {
+		Common::FSNode fileNode(getChildRecursive(_shaderPreset->basePath, i->fileName));
+
+		Common::SeekableReadStream *stream = fileNode.createReadStream();
+		if (!stream) {
+			warning("LibRetroPipeline::loadPasses: Could not open file '%s'", fileNode.getName().c_str());
```

Comment:
It will trigger an assert if the filenode is invalid. Maybe replace the `'fileNode.getName().c_str()` with `fileNode.isReadable() ? fileNode.getName().c_str() : i->fileName.c_str()`...

(test case: cel/presets/MMJ_Cel_Shader_3dfx.glslp)

Reply:
Good point, done